### PR TITLE
Improved header merge

### PR DESCRIPTION
**Pull request by daviesrob**
Builds on #419, extending the use of hash tables to all record types and removing all use of regular expression searches from bam_sort.c.  This should avoid quadratic behaviour when merging headers.

### CODE REVIEW

**Comment by daviesrob** [comment]
Some benchmarks.  First grab some data:

```
samtools view -H ftp://ftp.sra.ebi.ac.uk/vol1/ERA148/ERA148753/bam/7953_1#3.bam > 7953_1#3.sam
samtools view -H ftp://ftp.sra.ebi.ac.uk/vol1/ERA148/ERA148753/bam/7953_1#4.bam > 7953_1#4.sam
samtools view -H ftp://ftp.sra.ebi.ac.uk/vol1/ERA148/ERA148753/bam/7953_1#7.bam > 7953_1#7.sam
samtools view -H ftp://ftp.sra.ebi.ac.uk/vol1/ERA148/ERA148753/bam/7953_1#8.bam > 7953_1#8.sam
samtools view -H ftp://ftp.sra.ebi.ac.uk/vol1/ERA359/ERA359492/cram/13511_2#38.cram > 13511_2#38.sam
samtools view -H ftp://ftp.sra.ebi.ac.uk/vol1/ERA285/ERA285990/cram/11409_4.cram > 11409_4.sam
```

Pull request #436 (similar to current develop):
```
time samtools merge -u -f out.bam 13511_2#38.sam 7953_1#3.sam 7953_1#4.sam 7953_1#7.sam 7953_1#8.sam 11409_4.sam

real	904m25.953s
user	902m35.521s
sys 	0m7.956s
```

Pull request #419:
```
time samtools merge -u -f out.bam 13511_2#38.sam 7953_1#3.sam 7953_1#4.sam 7953_1#7.sam 7953_1#8.sam 11409_4.sam

real	0m2.796s
user	0m2.340s
sys 	0m0.224s
```
Much better, but changing the input file order has an interesting effect:
```
samtools merge -u -f out.bam 11409_4.sam 13511_2#38.sam 7953_1#3.sam 7953_1#4.sam 7953_1#7.sam 7953_1#8.sam

real	0m9.695s
user	0m9.117s
sys 	0m0.384s
```

With this pull request, the order makes no difference:
```
samtools merge -u -f out.bam 13511_2#38.sam 7953_1#3.sam 7953_1#4.sam 7953_1#7.sam 7953_1#8.sam 11409_4.sam

real	0m1.753s
user	0m1.372s
sys 	0m0.220s

samtools merge -u -f out.bam 11409_4.sam 13511_2#38.sam 7953_1#3.sam 7953_1#4.sam 7953_1#7.sam 7953_1#8.sam

real	0m1.791s
user	0m1.352s
sys 	0m0.248s

```

File: bam_sort.c [hunk 1]
```diff
@@ -140,376 +239,721 @@ static void trans_tbl_destroy(trans_tbl_t *tbl) {
...
+        // Take a copy of the ID as we'll need it for a hash key.
+        if (match_to_ks(text, &matches[1], &orig_id)) goto memfail;
+        
+        // is our matched ID in our output ID set already?
+        iter = kh_get(cset, known_ids, ks_str(&orig_id));
+        not_found_in_output = (iter == kh_end(known_ids));
+
+        if (override) {
+            // Override original ID (RG only)
+#ifdef OVERRIDE_DOES_NOT_MERGE
```
Comment (jmarshall) [inline bam_sort.c]: ???

Comment (daviesrob) [inline bam_sort.c]: This is the question of what to do when the -r option is in use (to force a read group for the input file) but the input file contains more than one read group.   At the moment they all end up with the same RG, which is similar to what it did before.  Defining OVERRIDE_DOES_NOT_MERGE makes it generate unique IDs for each RG instead, with override as  a prefix.  That is a change in behaviour, and also makes the tests fail, so some discussion is needed before we decide to go that way.  Hence it's there but not enabled at the moment.

Comment (jmarshall) [inline bam_sort.c]: Thanks.  Let's remember to bring it up for discussion soon!

**Pull request merged**